In [1]:
import os, sys
sys.path.append("..")

In [2]:
!python ../setup.py build 
!python ../setup.py build_ext --inplace

c:\Users\cmat411a\Documents\GitHub\Org_OptTrot\OptTrot\opttrot
running build
running build_py
running build_ext
building 'pauli_c' extension
"C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.39.33519\bin\HostX86\x64\cl.exe" /c /nologo /O2 /W3 /GL /DNDEBUG /MD -DBIG_NUM_BYTES=512 -Ic:\Users\cmat411a\Documents\GitHub\Org_OptTrot\OptTrot\opttrot -Ic:\Users\cmat411a\Documents\GitHub\Org_OptTrot\OptTrot\opttrot\c_src -Ic:\Users\cmat411a\Documents\GitHub\Org_OptTrot\OptTrot\opttrot\c_src\bn -Ic:\Users\cmat411a\anaconda3\envs\QC\include -Ic:\Users\cmat411a\anaconda3\envs\QC\Include "-IC:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.39.33519\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Auxiliary\VS\include" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\um" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.2

file c:\Users\cmat411a\Documents\GitHub\Org_OptTrot\OptTrot\opttrot\opttrot.py (for module opttrot) not found
file c:\Users\cmat411a\Documents\GitHub\Org_OptTrot\OptTrot\opttrot\opttrot.py (for module opttrot) not found


c:\Users\cmat411a\Documents\GitHub\Org_OptTrot\OptTrot\opttrot
running build_ext
copying build\lib.win-amd64-cpython-311\pauli_c.cp311-win_amd64.pyd -> c:\Users\cmat411a\Documents\GitHub\Org_OptTrot\OptTrot\opttrot


## PauliElement 

In [3]:
#import os, sys
#sys.path.append("..")

In [3]:
from opttrot.pauli import PauliElement

In [34]:
n = 3500
p00 = PauliElement(nx=3, nz= 6, n=n, weight=0.124, f=0, set_phase=False)
p01 = PauliElement(nx=5, nz= 5, n=n, weight=0.124, f=0, set_phase=False)


In [35]:
p00

PauliElement(n=3500, weight=0.124000, IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIZYX)

In [36]:
print(p00)

IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIZYX


## Speed comparsion

In [37]:
from numbers import Number
from functools import reduce
import numpy as np

## Compare with Qiskit

In [38]:
def to_qiskit(self):
    from qiskit.quantum_info import Pauli as qiskit_Pauli
    x_array = np.flip(np.array([int(b) for b in format(self.nx, f"0{self.n}b")], dtype=np.uint))
    z_array = np.flip(np.array([int(b) for b in format(self.nz, f"0{self.n}b")], dtype=np.uint))
    return self.weight, qiskit_Pauli((z_array, x_array))

In [39]:
p00 = PauliElement(nx=3, nz= 6, n=n, weight=0.124, f=0, set_phase=False)
p01 = PauliElement(nx=5, nz= 5, n=n, weight=0.124, f=0, set_phase=False)
w00, q_p00 = to_qiskit(p00)
w01, q_p01 = to_qiskit(p01)


In [40]:
q_p00.compose(q_p01)

Pauli('IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...')

In [41]:
p02= p00@p01
p02.pstr

'IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIXYZ'

## Speed

### Qiskit

In [42]:
%%timeit
q_p00.compose(q_p01)

32.8 µs ± 337 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


### OptTrot

In [43]:
%%timeit
p00@p01

1.35 µs ± 8.79 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


### Numpy binary

In [15]:
import numpy as np

nx1= np.random.randint(0, 2, size=n, dtype=bool)
nz1= np.random.randint(0, 2, size=n, dtype=bool)
nx2= np.random.randint(0, 2, size=n, dtype=bool)
nz2= np.random.randint(0, 2, size=n, dtype=bool)
f1 = 3
f2 = 4


In [16]:
%%timeit
nx3 = np.bitwise_xor(nx1, nx2)
nz3 = np.bitwise_xor(nz1, nz2)
f3 = f1+f2 + 2*np.bitwise_and(nx1, nz2).astype(int).sum() -  np.bitwise_and(nx3, nz3).astype(int).sum()

11.5 µs ± 36.2 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


### Result

- OS: Windows
- CPU: Ryzen 5 
- Python: 3.11.8


|Qiskit| OptTrot| Nb|
|:--:|:--:|:--:|
|27.9 µs ± 1.38 µs|1.35 µs ± 25.2 ns|11.4 µs ± 102 ns|

## PauliPoly

In [8]:
from opttrot.pauli import PauliElement, PauliPoly

In [9]:
p02 =p00@p01
PauliPoly([p00, p01, p02])

PauliPoly(terms:3)[
PauliElement(n=10, weight=0.124000, IIIIIIIZYX),
PauliElement(n=10, weight=0.124000, IIIIIIIYIY),
PauliElement(n=10, weight=0.015376, IIIIIIIXYZ)
]

In [10]:
hash(0.000000000000000000000001)

1528977390404266531

In [11]:
hash(p00)

-4352217537622120393

In [14]:
p_set = set([p00, p01, p02])

In [17]:
p03 = PauliElement(nx=1, nz= 1, n=n, weight=0.124, f=0, set_phase=False)
p04 = PauliElement(nx=10, nz= 1, n=n, weight=0.124, f=0, set_phase=False)
p_set2 = set([p00, p03, p04])

In [18]:
p_set & p_set2

{PauliElement(n=10, weight=0.124000, IIIIIIIZYX)}

In [20]:
p_set - p_set2

{PauliElement(n=10, weight=0.124000, IIIIIIIYIY),
 PauliElement(n=10, weight=0.015376, IIIIIIIXYZ)}

In [21]:
p_set.add(p00)

In [22]:
p00 in p_set

True

## Dict type Pauli Poly

In [ ]:
@classmethod
def from_pauli_list(cls, pauli_list:Iterable[PauliElement], coefs=None):
        if isinstance(coefs, Iterable) and not isinstance(coefs, str):
            for (p, coef) in zip(pauli_list, coefs):
                p.coef = coef
        p_dict = {}
        for p in pauli_list:
            p_dict[p.sym_code] = p
        return cls(p_dict, coefs)
@classmethod
def from_pauli_str_list(cls, pauli_str_list:Iterable[str], coefs=None):
        #if isinstance(coefs, Iterable) and not isinstance(coefs, str):
        #    for (p, coef) in zip(pauli_list, coefs):
        #        p.coef = coef
        p_dict = {}
        n = len(pauli_str_list[0])
        for pstr in pauli_str_list:
            nx, nz = pstr2sym_code(pstr)
            p_dict[(nx, nz)] = PauliElement(nx, nz, n, weight = 1)
        
        return cls(p_dict)